proposals/0621-linear-constraints.rst

```diff
@@ -0,0 +1,1108 @@
+consuming the type class once.
+
+We'd need a way to specify unrestricted methods, we could use the same
+syntax as records in the `Linear Types proposal`_
+
+::
+
+   class C a where
+     lin :: a -> T
+     unr %Many :: a -> U
+
+But for this relaxed condition to be useful we would need to be able
+to call ``unr`` on a linear instance of ``C`` *without consuming it*.
+
+In other words, we would need something like:
+
+::
+
+   lin :: C a %1 => a -> T
+   unr :: C a %0 => a -> U
+
+The calls to ``unr`` are free, they don't count toward the
+exactly-once consumption of the instance. This ``0`` wouldn't mean
+“erased at runtime” as has sometimes been proposed. Arnaud is pretty
+convinced that this particular ``0`` is a desirable feature for Linear
+Haskell, but there's little telling what kind of havoc it would wreak
+on multiplicity inference in its current state. So we'd rather keep
+this extension for a later time.
+
+Dupable classes
+^^^^^^^^^^^^^^^
+
+We specified dupable type classes in a way that make all dupable type
+classes isomorphic. The reason for the design is that what we actually
+have in mind in the long run is to be able to add unrestricted methods
+to the type class as in the alternative above. This design has the
+advantage that it's reasonably easy to implement, only wiring in a
+type (``LinearityToken```) and two functions
+(``consumeLinearityToken`` and ``dup2LinearityToken``). Nevertheless
+there are a number of other ways to go about dupable classes.
+
+Leave it out
+~~~~~~~~~~~~
+
+Let us point out that while the ability to form a ``Linearly``
+constraint is both quite useful and absolutely at home in this
+proposal, it's also perfectly consistent to make a linear constraints
+proposal without any dupable type classes. If the design of this
+feature proves too controversial, it's definitely an option to simply
+excise the dupable class feature from the proposal.
+
+Kind of LinearityToken
+~~~~~~~~~~~~~~~~~~~~~~
+
+The proposal deliberately specifies that ``LinearityToken`` be 0
+width, and the ``dup2`` and ``consume`` functions to correspondingly
+return unboxed tuples. This is meant to emphasise that this is all for
+low level manipulation and making sure that there is no cost in
+storing ``LinearityToken``. These are all meant to be used while
+defining a dupable type class and its API, but it's not intended for
+them to be apparent in said API.
+
+An somewhat middle-ground option is to expose
+
+::
+
+  data LinearityToken# :: ZeroBitType
+
+  data LinearityToken = MkLinearityToken LinearlyToken#
+
+  -- Both functions below can be defined as easily inlineable thin
+  -- wrapper so that in most cases no allocation is needed
+  consumeLinearityToken :: LinearityToken -> ()
+  dup2LinearityToken :: LinearityToken -> (LinearityToken, LinearityToken)
+
+Just linearly
+~~~~~~~~~~~~~
+
+The most useful dupable type class in the proximate future (in fact
+the only known example yet), is ``Linearly``. So another option, to
```

Comment:
To be clear, the fact that all dupable type classes are isomorphic means it's the only possible example, doesn't it?